unity/Assets/Scripts/FifoServer.cs

```diff
@@ -0,0 +1,126 @@
+        private byte[] eomHeader;
+
+        
+        public static Dictionary<string, FieldType> FormMap = new Dictionary<string, FieldType>{
+            {"image", FieldType.RGBImage},
+            {"image_depth", FieldType.DepthImage},
+            {"image_normals", FieldType.NormalsImage},
+            {"image_classes", FieldType.ClassesImage},
+            {"image_flow", FieldType.FlowsImage},
+            {"image_ids", FieldType.IDsImage},
+            {"image-thirdParty-camera", FieldType.ThirdPartyCameraImage}
+
+        };
+
+        static public int UnpackNetworkBytes(byte[] data, int offset=0) {
+            int networkInt = System.BitConverter.ToInt32(data, offset);
+            return IPAddress.NetworkToHostOrder(networkInt);
+        }
+
+        static public byte[] PackNetworkBytes(int val) {
+            int networkInt = IPAddress.HostToNetworkOrder(val);
+            return System.BitConverter.GetBytes(networkInt);
+        }
+
+        public string ReceiveMessage() {
+            if (clientPipe == null) {
+                clientPipe = new NamedPipeClientStream(".", this.clientPipePath, PipeDirection.In);
+                clientPipe.Connect();
+            }
+            string action = null;
+            while (true) {
+                byte[] header = new byte[headerLength];
+                clientPipe.Read(header, 0, header.Length);
+                FieldType fieldType = (FieldType)header[0];
+                if (fieldType == FieldType.EndOfMessage) {
+                        //Console.WriteLine("Got eom");
+                    break;
+                }
+                int fieldLength = UnpackNetworkBytes(header, 1);
+                byte[] body = new byte[fieldLength];
+                clientPipe.Read(body, 0, body.Length);
+
+                switch (fieldType) {
+                    case FieldType.Action:
+                        //Console.WriteLine("Got action");
+                        action = Encoding.Default.GetString(body);
+                        break;
+                    default:
+                        throw new Exception("Invalid field type for Client: " + fieldType);
+                }
+            }
+            return action;
+        }
+
+        public void SendEOM() {
+            serverPipe.Write(this.eomHeader, 0, this.headerLength);
+            serverPipe.Flush();
+        }
+
+        public void SendMessage(FieldType t, byte[] body) {
+            if (this.serverPipe == null) {
+                this.serverPipe = new NamedPipeClientStream(".", this.serverPipePath, PipeDirection.Out);
+                this.serverPipe.Connect();
+            }
+            //Console.WriteLine("server pipe + connected " + this.serverPipe.IsConnected );
+            byte[] header = new byte[headerLength];
+            header[0] = (byte)t;
+            PackNetworkBytes(body.Length).CopyTo(header, 1);
+            serverPipe.Write(header, 0, header.Length);
+            serverPipe.Write(body, 0, body.Length);
```

Comment:
Does making 2 `Write` calls vs one change anything? the server is still waiting at this point right, so it's just for clarity?

Reply:
1 write call vs 2 really showed no difference - you would also have to create a new byte[] array and copy the data from each into the new one in order to get the single write call, so you lose any time you save doing that.  As well, serverPipe buffers the write calls internally and flushes, so it ended up not making any difference.

Reply:
makes sense